### ISSUE TITLE
[Feature Request] Optifine Uninstall Option through the Installer

### ISSUE THREAD

**Issue by HiItsDevin** [sp614x/optifine#5449]
Have an option in the Optifine installer to be able to uninstall the mod, in the scenario that someone no longer wants to use OptiFine, and doesn't want to go through the effort of trying to find every OptiFine file in the .minecraft directory, to delete the mod.

This is probably really dumb but it could be helpful, especially in the case of a bad/corrupt install too.

Quick mockup of what it could look like [done in WinForms]

![image](https://user-images.githubusercontent.com/30705254/111255118-ae12e680-85d3-11eb-80b8-2ed04c56d6ac.png)

**Comment by LizAinslie** [comment]
i +1 this. personally no reason for me to uninstall but it could be handy in the event that someone no longer needs or wants optifine

**Comment by UnknownTheLad** [comment]
Yeah this would be cool, would prefer keeping the original window size though. Other than that I have no complaints! :D

**Comment by HiItsDevin** [comment]
oh i only stretched out the size so it was easier for me to make in Visual Studio, my bad, but yes, keeping the original size, or just expanding it a little would be preferred. c:

**Comment by Regnander** [comment]
Can you add "Reinstall" as an option to the mockup as well, or do you think it would be too messy with an extra button?

**Comment by KaiAF** [comment]
Forgot to actually give my opinion on this last night. I believe this should be added, there are people who have corrupted installs, and they need to manually delete the folders (If they are there). Sometimes they end up deleting other folders/files.

**Comment by HiItsDevin** [comment]
> Can you add "Reinstall" as an option to the mockup as well, or do you think it would be too messy with an extra button?

It would probably be too big to fit the original installer, but sure, I can make a mock-up regardless.

**Comment by HiItsDevin** [comment]
> Forgot to actually give my opinion on this last night. I believe this should be added, there are people who have corrupted installs, and they need to manually delete the folders (If they are there). Sometimes they end up deleting other folders/files. 

Yeah I've seen a lot of that in the support channels on the discord, I do believe this would really help quite a bit.

**Comment by sp614x** [comment]
You can just delete the profile in the launcher.
Installing a second time will recreate the profile.

**Comment by KaiAF** [comment]
What about the folder in `.minecraft\libraries\optifine\OptiFine`? I know from the support channel that sometimes people don't feel comfortable manually deleting something in the versions folder or the libraries folder. We usually have to explain that it wouldn't affect much. 
Sometimes there are people who just want to fully delete all optifine related folders (Can't tell you why). And usually, they don't feel comfortable manually going in and deleting folders. 

From what I can remember, this is the process we usually have to go through. (Could be a bit wrong because personally, I haven't seen it in a bit)

1. User comes to the channel saying that they installed optifine but it will not show in the versions folder at all. We confirm it by asking for them to provide a screenshot (Sometimes they provide a video).
2. We tell them to re-install optifine via the installer. The same issue happens again.
3. We tell them to go to `.minecraft\versions` and see if it shows up there. Sometimes it does, sometimes it doesn't.
4. We direct them to `.minecraft\libraries\optifine\OptiFine` and ask them if they see the OptiFine version name as a folder. They usually do. We tell them to delete it. And re-install. Usually fixes it.

With an uninstall button, this could prevent *all* of that. Could even prevent people from coming to the support channel.

**Comment by HiItsDevin** [comment]
Exactly this, it would prove really helpful ^

**Comment by LeotomasMC** [comment]
> Can you add "Reinstall" as an option to the mockup as well, or do you think it would be too messy with an extra button?

If you want to reinstall, cant you just run the installer again as if you had never installed it in the first place though? I don't see any benefit of having a button specifically to *re*-install when it just does what the normal install button does.

**Comment by Regnander** [comment]
> If you want to reinstall, cant you just run the installer again as if you had never installed it in the first place though? I don't see any benefit of having a button specifically to _re_-install when it just does what the normal install button does.

In some installers, you get the option to repair any existing installation. I should have just said "Repair" instead of "Reinstall" in the first place since that's what I had in mind.

![](https://user-images.githubusercontent.com/17113053/111524318-8e5afa00-875c-11eb-9563-304b124c9f5a.PNG)

**Comment by HiItsDevin** [comment]
> > If you want to reinstall, cant you just run the installer again as if you had never installed it in the first place though? I don't see any benefit of having a button specifically to _re_-install when it just does what the normal install button does.
> 
> In some installers, you get the option to repair any existing installation. I should have just said "Repair" instead of "Reinstall" in the first place since that's what I had in mind.
> 
> ![](https://user-images.githubusercontent.com/17113053/111524318-8e5afa00-875c-11eb-9563-304b124c9f5a.PNG)

Like this?

![image](https://user-images.githubusercontent.com/30705254/111525384-cc372e80-871a-11eb-87c6-3b28dbf138d9.png)

**Comment by Regnander** [comment]
> Like this?
> 
> ![image](https://user-images.githubusercontent.com/30705254/111525384-cc372e80-871a-11eb-87c6-3b28dbf138d9.png)

Hmm, now I'm having second thoughts since I don't want the users - especially the younger Minecraft players - to get overwhelmed by buttons everywhere and scare them off.

On top of my head I can think of two solutions for this:

* Replace the "Install" button with "Repair" if an OptiFine version already is installed.
  * Pros: Less buttons, slightly easier to use.
  * Cons: Might be confusing if a version gets hotfixed and the user is trying to look for the "Install" button.
* Disable the "Repair" button if an OptiFine version is already installed.
  * Pros: Fixes the cons mentioned above.
  * Cons: More buttons.

**Comment by sp614x** [comment]
The installer can only repair the installation of its own version which is exactly the same as installing it.
There are tons of libraries in the folder "libraries". The Mojang launcher automatically downloads the libraries and never deletes them. It also downloads all resources for every MC version in the folder "assets" and never deletes them. 
The Mojang launcher doesn't uninstall anything, ever. Why would then the OptiFine versions in "libraries" be a concern?

**Comment by sp614x** [comment]
Another thing is, should the uninstaller remove only the version it knows about or should it remove all OptiFine versions?
If all versions are to be uninstalled, then it would probably have to delete all OptiFine launcher profiles as well.
An button Cleanup could be used to delete all OptiFine libraries and versions which are not used in any launcher profile, not sure how useful it would be.

**Comment by KaiAF** [comment]
> Another thing is, should the uninstaller remove only the version it knows about or should it remove all OptiFine versions?

If possible, remove the *one* version. But I don't know if it can. 

> A button Cleanup could be used to delete all OptiFine libraries and versions which are not used in any launcher profile, not sure how useful it would be.

I mean, this could help with the "corrupted installs" as I mentioned before, where they would have to delete the specific folder in libraries. It's only useful as it's easier to explain to someone to just press a button on the installer, rather than going through all the steps said before. Possibly the user would do that first before even going to support, which is a plus to have. No idea.

**Comment by sp614x** [comment]
The installer should fix the corrupted installs as it sets everything as it should be for the installation to run.

**Comment by HiItsDevin** [comment]
If I recall correctly, it doesn't sometimes
